Crash while building

While building on my computer I get the following error:

```
[ 22%] Building CXX object src/manifold/CMakeFiles/manifold.dir/src/boolean_result.cpp.o
cd /app/manifold/build/src/manifold && /usr/bin/c++ -Dmanifold_EXPORTS -I/app/manifold/src/manifold/include -I/app/manifold/src/utilities/include -I/app/manifold/src/third_party/thrust/dependencies/libcudacxx/include -I/app/manifold/src/third_party/thrust -I/app/manifold/src/third_party/glm/glm/.. -I/app/manifold/src/cross_section/include -I/app/manifold/src/third_party/clipper2/CPP/Clipper2Lib/include -I/app/manifold/src/collider/include -I/app/manifold/src/polygon/include -I/app/manifold/src/third_party/graphlite/include -O3 -DNDEBUG -fPIC -fPIC -Werror -Wall -Wno-sign-compare -Wno-unused -O3 -DTHRUST_DEVICE_SYSTEM=THRUST_DEVICE_SYSTEM_CPP -std=gnu++17 -MD -MT src/manifold/CMakeFiles/manifold.dir/src/boolean_result.cpp.o -MF CMakeFiles/manifold.dir/src/boolean_result.cpp.o.d -o CMakeFiles/manifold.dir/src/boolean_result.cpp.o -c /app/manifold/src/manifold/src/boolean_result.cpp
In file included from /app/manifold/src/utilities/include/sparse.h:22,
                 from /app/manifold/src/collider/include/collider.h:17,
                 from /app/manifold/src/manifold/src/impl.h:18,
                 from /app/manifold/src/manifold/src/boolean3.h:16,
                 from /app/manifold/src/manifold/src/boolean_result.cpp:18:
In static member function ‘static void manifold::ManagedVec<T>::mallocManaged(T**, size_t) [with T = int]’,
    inlined from ‘std::tuple<manifold::VecDH<int>, manifold::VecDH<int> > {anonymous}::SizeOutput(manifold::Manifold::Impl&, const manifold::Manifold::Impl&, const manifold::Manifold::Impl&, const manifold::VecDH<int>&, const manifold::VecDH<int>&, const manifold::VecDH<int>&, const manifold::VecDH<int>&, const manifold::SparseIndices&, const manifold::SparseIndices&, bool, manifold::ExecutionPolicy)’ at /app/manifold/src/utilities/include/vec_dh.h:155:20:
/app/manifold/src/utilities/include/vec_dh.h:261:42: error: argument 1 value ‘18446744073709551612’ exceeds maximum object size 9223372036854775807 [-Werror=alloc-size-larger-than=]
       *ptr = reinterpret_cast<T *>(malloc(bytes));
                                    ~~~~~~^~~~~~~
In file included from /usr/include/c++/8/cstdlib:75,
                 from /usr/include/c++/8/bits/stl_algo.h:59,
                 from /usr/include/c++/8/algorithm:62,
                 from /app/manifold/src/manifold/src/boolean_result.cpp:15:
/usr/include/stdlib.h: In function ‘std::tuple<manifold::VecDH<int>, manifold::VecDH<int> > {anonymous}::SizeOutput(manifold::Manifold::Impl&, const manifold::Manifold::Impl&, const manifold::Manifold::Impl&, const manifold::VecDH<int>&, const manifold::VecDH<int>&, const manifold::VecDH<int>&, const manifold::VecDH<int>&, const manifold::SparseIndices&, const manifold::SparseIndices&, bool, manifold::ExecutionPolicy)’:
/usr/include/stdlib.h:539:14: note: in a call to allocation function ‘void* malloc(size_t)’ declared here
 extern void *malloc (size_t __size) __THROW __attribute_malloc__ __wur;
              ^~~~~~
cc1plus: all warnings being treated as errors
src/manifold/CMakeFiles/manifold.dir/build.make:92: recipe for target 'src/manifold/CMakeFiles/manifold.dir/src/boolean_result.cpp.o' failed
make[2]: *** [src/manifold/CMakeFiles/manifold.dir/src/boolean_result.cpp.o] Error 1
make[2]: Leaving directory '/app/manifold/build'
CMakeFiles/Makefile2:668: recipe for target 'src/manifold/CMakeFiles/manifold.dir/all' failed
make[1]: *** [src/manifold/CMakeFiles/manifold.dir/all] Error 2
make[1]: Leaving directory '/app/manifold/build'
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
```

Any idea what might cause this?

Hi, it was indeed a fresh clone from the repo.
On your branch I'm getting the same error, now only at 24% completion. But it seems to be in the same file.

```
[ 24%] Building CXX object src/manifold/CMakeFiles/manifold.dir/src/boolean_result.cpp.o
cd /app/manifold/build/src/manifold && /usr/bin/c++ -Dmanifold_EXPORTS -I/app/manifold/src/manifold/include -I/app/manifold/src/utilities/include -I/app/manifold/src/third_party/thrust/dependencies/libcudacxx/include -I/app/manifold/src/third_party/thrust -I/app/manifold/src/third_party/glm/glm/.. -I/app/manifold/src/cross_section/include -I/app/manifold/src/third_party/clipper2/CPP/Clipper2Lib/include -I/app/manifold/src/collider/include -I/app/manifold/src/polygon/include -I/app/manifold/src/third_party/graphlite/include -O3 -DNDEBUG -fPIC -fPIC -Werror -Wall -Wno-sign-compare -Wno-unused -O3 -DTHRUST_DEVICE_SYSTEM=THRUST_DEVICE_SYSTEM_CPP -std=gnu++17 -MD -MT src/manifold/CMakeFiles/manifold.dir/src/boolean_result.cpp.o -MF CMakeFiles/manifold.dir/src/boolean_result.cpp.o.d -o CMakeFiles/manifold.dir/src/boolean_result.cpp.o -c /app/manifold/src/manifold/src/boolean_result.cpp
In file included from /app/manifold/src/utilities/include/sparse.h:22,
                 from /app/manifold/src/collider/include/collider.h:17,
                 from /app/manifold/src/manifold/src/impl.h:18,
                 from /app/manifold/src/manifold/src/boolean3.h:16,
                 from /app/manifold/src/manifold/src/boolean_result.cpp:18:
In static member function ‘static void manifold::ManagedVec<T>::mallocManaged(T**, size_t) [with T = int]’,
    inlined from ‘std::tuple<manifold::VecDH<int>, manifold::VecDH<int> > {anonymous}::SizeOutput(manifold::Manifold::Impl&, const manifold::Manifold::Impl&, const manifold::Manifold::Impl&, const manifold::VecDH<int>&, const manifold::VecDH<int>&, const manifold::VecDH<int>&, const manifold::VecDH<int>&, const manifold::SparseIndices&, const manifold::SparseIndices&, bool, manifold::ExecutionPolicy)’ at /app/manifold/src/utilities/include/vec_dh.h:155:20:
/app/manifold/src/utilities/include/vec_dh.h:261:42: error: argument 1 value ‘18446744073709551612’ exceeds maximum object size 9223372036854775807 [-Werror=alloc-size-larger-than=]
       *ptr = reinterpret_cast<T *>(malloc(bytes));
                                    ~~~~~~^~~~~~~
In file included from /usr/include/c++/8/cstdlib:75,
                 from /usr/include/c++/8/bits/stl_algo.h:59,
                 from /usr/include/c++/8/algorithm:62,
                 from /app/manifold/src/manifold/src/boolean_result.cpp:15:
/usr/include/stdlib.h: In function ‘std::tuple<manifold::VecDH<int>, manifold::VecDH<int> > {anonymous}::SizeOutput(manifold::Manifold::Impl&, const manifold::Manifold::Impl&, const manifold::Manifold::Impl&, const manifold::VecDH<int>&, const manifold::VecDH<int>&, const manifold::VecDH<int>&, const manifold::VecDH<int>&, const manifold::SparseIndices&, const manifold::SparseIndices&, bool, manifold::ExecutionPolicy)’:
/usr/include/stdlib.h:539:14: note: in a call to allocation function ‘void* malloc(size_t)’ declared here
 extern void *malloc (size_t __size) __THROW __attribute_malloc__ __wur;
              ^~~~~~
cc1plus: all warnings being treated as errors
src/manifold/CMakeFiles/manifold.dir/build.make:92: recipe for target 'src/manifold/CMakeFiles/manifold.dir/src/boolean_result.cpp.o' failed
make[2]: *** [src/manifold/CMakeFiles/manifold.dir/src/boolean_result.cpp.o] Error 1
make[2]: Leaving directory '/app/manifold/build'
CMakeFiles/Makefile2:668: recipe for target 'src/manifold/CMakeFiles/manifold.dir/all' failed
make[1]: *** [src/manifold/CMakeFiles/manifold.dir/all] Error 2
make[1]: Leaving directory '/app/manifold/build'
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
```

--- a/flake.nix
+++ b/flake.nix
@@ -16,7 +16,7 @@
             , doCheck ? true
             , build-tools ? [ ]
             , ...
-            }: pkgs.llvmPackages_15.stdenv.mkDerivation {
+            }: pkgs.stdenv.mkDerivation {
               inherit doCheck;
               pname =
                 if cuda-support then

--- a/src/utilities/include/vec_dh.h
+++ b/src/utilities/include/vec_dh.h
@@ -228,19 +228,9 @@ class ManagedVec {
 
   const T &back() const { return *(ptr_ + size_ - 1); }
 
-  T &operator[](size_t i) {
-    // ASSERT(i < size_, logicErr,
-    //        "Index out of bounds: " + std::to_string(i) +
-    //            ", size: " + std::to_string(size_));
-    return *(ptr_ + i);
-  }
+  T &operator[](size_t i) { return *(ptr_ + i); }
 
-  const T &operator[](size_t i) const {
-    // ASSERT(i < size_, logicErr,
-    //        "Index out of bounds: " + std::to_string(i) +
-    //            ", size: " + std::to_string(size_));
-    return *(ptr_ + i);
-  }
+  const T &operator[](size_t i) const { return *(ptr_ + i); }
 
   bool empty() const { return size_ == 0; }
 
@@ -253,6 +243,10 @@ class ManagedVec {
   static constexpr int DEVICE_MAX_BYTES = 1 << 16;
 
   static void mallocManaged(T **ptr, size_t bytes) {
+    // only exists to please the compiler
+    if (bytes >= (1ull << 63)) {
+      throw std::bad_alloc();
+    }
 #ifdef MANIFOLD_USE_CUDA
     if (CudaEnabled())
       cudaMallocManaged(reinterpret_cast<void **>(ptr), bytes);

--- a/src/utilities/include/vec_dh.h
+++ b/src/utilities/include/vec_dh.h
@@ -244,6 +244,7 @@ class ManagedVec {
 
   static void mallocManaged(T **ptr, size_t bytes) {
     // only exists to please the compiler
+    // see https://github.com/elalish/manifold/pull/496 for background
     if (bytes >= (1ull << 63)) {
       throw std::bad_alloc();
     }

